userspace/libsinsp/thread_pool.h

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+Copyright (C) 2024 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#include <list>
+#include <cstdint>
+#include <cstddef>
+#include <functional>
+#include <memory>
+
+class thread_pool
+{
+public:
+	using routine_id_t = std::function<bool()>*;
+
+	thread_pool() = default;
+
+	virtual ~thread_pool() = default;
+
+	/*!
+  	* \brief Subscribes a routine to the thread pool.
+	*
+	* \param func The routine to be subscribed, represented by a function returning a bool value.
+	* Returning false causes the routine to be unsubscribed from the thread pool.
+	*
+	* \return An handle representing a specific routine.
+	* This can later be used to unsubscribe the routine.
+	*/
+	virtual routine_id_t subscribe(const std::function<bool()>& func) = 0;
+
+	/*!
+  	* \brief Unsubscribes a routine from the thread pool.
+	*
+	* \param id A routine handle.
+	*/
+	virtual void unsubscribe(routine_id_t id) = 0;
+
+	/*!
+  	* \brief Unsubscribes all the subscribed routines and waits for the running ones to finish.
+	*/
+	virtual void purge() = 0;
+
+	/*!
+	* \return The count of currently subscribed routines.
+	*/
+	virtual size_t routines_num() = 0;
+};
+
+namespace BS {
+	class thread_pool;
+};
+
+class bs_thread_pool : public thread_pool
```

Comment:
I'd split this into its own separate header file `thread_pool_bs.h`. The same goes for the current `thread_pool.cpp`, which might become `thread_pool_bs.cpp`. This will allow people to not include an header of implementations they are not interested into.